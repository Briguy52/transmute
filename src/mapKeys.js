import curry from "./curry";
import { Collection, Seq } from "immutable";
import { mapKeys } from "./protocols/Keyed";

mapKeys.implement(Object, (keyMapper, subject) => {
  const keys = Object.keys(subject);
  const len = keys.length;
  const result = {};
  for (let i = 0; i < len; i++) {
    const key = keys[i];
    const value = subject[key];
    result[keyMapper(key, value, subject)] = value;
  }
  return result;
});

mapKeys.implementInherited(Collection.Keyed, (keyMapper, subject) => {
  const nextEntries = subject
    .entrySeq()
    .map(([key, value]) => [keyMapper(key, value, subject), value]);
  return new subject.constructor(nextEntries);
});

mapKeys.implementInherited(Seq.Keyed, (keyMapper, subject) => {
  const nextEntries = subject
    .entrySeq()
    .map(([key, value]) => [keyMapper(key, value, subject), value]);
  return Seq.Keyed(nextEntries);
});

/**
 * Like `map` but transforms an Iterable's keys rather than its values.
 *
 * @example <caption>Can be useful for converting keys of API results to a common type.</caption>
 * import { mapKeys, toString } from 'transmute';
 * const stringifyKeys = mapKeys(toString);
 * const m = Map.of(123, Map(), 456, Map(), 789, Map());
 * stringifyKeys(m).equals(Map.of('123', Map(), '456', Map(), '789', Map()));
 *
 * @param  {Function} keyMapper returns a new key
 * @param  {KeyedIterable} subject
 * @return {KeyedIterable}
 */
export default curry(mapKeys);
