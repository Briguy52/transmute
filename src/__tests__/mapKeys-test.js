import { List, Map, OrderedMap, Seq, Set } from "immutable";
import mapKeys from "../mapKeys";

describe("transmute/mapKeys", () => {
  const prependTest = mapKeys(key => `test-${key}`);

  it("passes the key, value, and iterable to keyMapper", () => {
    const iterable = Map({ one: 1 });
    const keyMapper = jest.fn();
    mapKeys(keyMapper, iterable);
    expect(keyMapper).toHaveBeenCalledWith("one", 1, iterable);
  });

  it("throws for non-Keyed Iterables", () => {
    expect(() => prependTest(List())).toThrow();
    expect(() => prependTest(Seq([]))).toThrow();
    expect(() => prependTest(Set())).toThrow();
  });

  it("maps Map keys", () => {
    const original = Map([[1, 1], [2, 2], [3, 3]]);
    expect(prependTest(original)).toMatchSnapshot();
  });

  it("maps Objects", () => {
    const original = { 1: 1, 2: 2, 3: 3 };
    expect(prependTest(original)).toMatchSnapshot();
  });

  it("maps OrderedMap keys", () => {
    const original = OrderedMap([[1, 1], [2, 2], [3, 3]]);
    expect(prependTest(original)).toMatchSnapshot();
  });
});
