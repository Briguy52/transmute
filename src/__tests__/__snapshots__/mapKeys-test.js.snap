// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/mapKeys maps Map keys 1`] = `
Immutable.Map {
  test-1: 1,
  test-2: 2,
  test-3: 3,
}
`;

exports[`transmute/mapKeys maps Objects 1`] = `
Object {
  "test-1": 1,
  "test-2": 2,
  "test-3": 3,
}
`;

exports[`transmute/mapKeys maps OrderedMap keys 1`] = `
Immutable.OrderedMap {
  test-1: 1,
  test-2: 2,
  test-3: 3,
}
`;
